File: source/mcu/src/tasks/control.c
#include "control.h"

#include <hal/assert.h>
#include <hal/math.h>

#define RB_STRUCT DacRingBuffer
#define RB_PREFIX dac_rb
#define RB_ITEM point_t
#define RB_CAPACITY DAC_BUFFER_SIZE
#include <utils/ringbuf.inl>
#undef RB_STRUCT
#undef RB_PREFIX
#undef RB_ITEM
#undef RB_CAPACITY

#define RB_STRUCT AdcRingBuffer
#define RB_PREFIX adc_rb
#define RB_ITEM AdcArray
#define RB_CAPACITY ADC_BUFFER_SIZE
#include <utils/ringbuf.inl>
#undef RB_STRUCT
#undef RB_PREFIX
#undef RB_ITEM
#undef RB_CAPACITY

void control_init(Control *self, Statistics *stats) {
    self->dio.in = 0;
    self->dio.out = 0;

    self->dac.running = false;
    hal_assert_retcode(dac_rb_init(&self->dac.buffer));
    self->dac.last_point = 0x7fff;
    self->dac.counter = 0;

    hal_assert_retcode(adc_rb_init(&self->adc.buffer));
    self->adc.counter = 0;

    self->sync = NULL;

    hal_assert(stats != NULL);
    self->stats = stats;
}

void control_deinit(Control *self) {
    hal_assert_retcode(dac_rb_deinit(&self->dac.buffer));
    hal_assert_retcode(adc_rb_deinit(&self->adc.buffer));
}

void control_set_sync(Control *self, ControlSync *sync) {
    hal_assert(sync != NULL);
    hal_assert(sync->ready_sem != NULL);
    self->sync = sync;
}

void control_dac_start(Control *self) {
    skifio_dac_enable();
    self->dac.running = true;
}

void control_dac_stop(Control *self) {
    self->dac.running = false;
    skifio_dac_disable();
}

void control_sync_init(ControlSync *self, SemaphoreHandle_t *ready_sem, size_t dac_chunk_size, size_t adc_chunk_size) {
    self->ready_sem = ready_sem;

    self->dac_notify_every = dac_chunk_size;
    self->adc_notify_every = adc_chunk_size;

    self->din_changed = false;
    self->dout_changed = false;
}

static bool update_din(Control *self) {
    SkifioDin din = skifio_din_read();
    if (din != self->dio.in) {
        self->dio.in = din;
        self->sync->din_changed = true;
        return true;
    } else {
        return false;
    }
}

static void intr_din_handler(void *data, SkifioDin value) {
    Control *self = (Control *)self;
    update_din(self);
    BaseType_t hptw = pdFALSE;
    if (self->sync->din_changed) {
        xSemaphoreGiveFromISR(*self->sync->ready_sem, &hptw);
    }
    portYIELD_FROM_ISR(hptw);
}

static void control_task(void *param) {
    Control *self = (Control *)param;

    // Check that sync is set.
    hal_assert(self->sync != NULL);

    hal_log_info("SkifIO driver init");
    hal_assert_retcode(skifio_init());
    hal_assert_retcode(skifio_din_subscribe(intr_din_handler, NULL));

    hal_log_info("Enter SkifIO loop");
    uint64_t prev_intr_count = _SKIFIO_DEBUG_INFO.intr_count;
    for (size_t k = 0;; ++k) {
        bool ready = false;

        // Wait for 10 kHz sync signal
        {
            hal_retcode ret = skifio_wait_ready(1000);
            if (ret == HAL_TIMED_OUT) {
                hal_log_warn("SkifIO timeout %d", k);
                continue;
            }
            hal_assert_retcode(ret);
        }

        // Write discrete output
        if (self->sync->dout_changed) {
            hal_assert_retcode(skifio_dout_write(self->dio.out));
            self->sync->dout_changed = false;
        }

        // Read discrete input
        ready |= update_din(self);

        // Statistics: detect 10 kHz sync signal loss
        self->stats->max_intrs_per_sample = hal_max(
            self->stats->max_intrs_per_sample,
            (uint32_t)(_SKIFIO_DEBUG_INFO.intr_count - prev_intr_count) //
        );
        prev_intr_count = _SKIFIO_DEBUG_INFO.intr_count;

        // Fetch next DAC value from buffer
        int32_t dac_value = self->dac.last_point;
        if (self->dac.running) {
            if (dac_rb_read(&self->dac.buffer, &dac_value, 1) == 1) {
                self->dac.last_point = dac_value;
                // Decrement DAC notification counter.
                if (self->dac.counter > 0) {
                    self->dac.counter -= 1;
                } else {
                    self->dac.counter = self->sync->dac_notify_every - 1;
                    ready = true;
                }
            } else {
                self->stats->dac.lost_empty += 1;
            }
        }

        // Transfer DAC/ADC values to/from SkifIO board.
        {
            SkifioInput input = {{0}};
            SkifioOutput output = {0};

            output.dac = (int16_t)dac_value;
            hal_retcode ret = skifio_transfer(&output, &input);
            if (ret == HAL_INVALID_DATA) {
                // CRC check error
                self->stats->crc_error_count += 1;
                ret = HAL_SUCCESS;
            }
            hal_assert_retcode(ret);

            // Handle ADCs
            AdcArray adcs = {{0}};
            for (size_t i = 0; i < ADC_COUNT; ++i) {
                point_t value = input.adcs[i];
                adcs.points[i] = value;

                // Update ADC value statistics
                value_stats_update(&self->stats->adc.values[i], value);
            }
            // Push ADC point to buffer.
            if (adc_rb_write(&self->adc.buffer, &adcs, 1) != 1) {
                self->stats->adc.lost_full += 1;
            }

            // Decrement ADC notification counter.
            if (self->adc.counter > 0) {
                self->adc.counter -= 1;
            } else {
                self->adc.counter = self->sync->adc_notify_every - 1;
                ready = true;
            }
        }

        if (ready) {
            // Notify
            xSemaphoreGive(*self->sync->ready_sem);
        }

        self->stats->sample_count += 1;
    }

    // This task must never end.
    hal_unreachable();

    hal_assert_retcode(skifio_deinit());
}

void control_run(Control *self) {
    hal_assert(xTaskCreate(control_task, "control", TASK_STACK_SIZE, (void *)self, CONTROL_TASK_PRIORITY, NULL) == pdPASS);
}

File: source/mcu/user/src/control.rs
use crate::{
    hal::RetCode,
    println,
    skifio::{self, Aout, AtomicDin, AtomicDout, Din, DinHandler, XferIn, XferOut},
    Error,
};
use alloc::sync::Arc;
use common::config::{Point, ADC_COUNT};
use core::{
    sync::atomic::{AtomicBool, AtomicU64, Ordering},
    time::Duration,
};
use freertos::{Semaphore, Task, TaskPriority};
use ringbuf::{StaticConsumer, StaticProducer};

pub type DacPoint = Point;
pub type AdcPoint = [Point; ADC_COUNT];

pub const DAC_BUFFER_LEN: usize = 1024;
pub const ADC_BUFFER_LEN: usize = 384;

pub type DacProducer = StaticProducer<'static, DacPoint, DAC_BUFFER_LEN>;
pub type DacConsumer = StaticConsumer<'static, DacPoint, DAC_BUFFER_LEN>;

pub type AdcProducer = StaticProducer<'static, AdcPoint, ADC_BUFFER_LEN>;
pub type AdcConsumer = StaticConsumer<'static, AdcPoint, ADC_BUFFER_LEN>;

pub struct StatsDac {
    pub lost_empty: AtomicU64,
}
pub struct StatsAdc {
    pub lost_full: AtomicU64,
}
pub struct Statistics {
    pub sample_count: AtomicU64,
    pub dac: StatsDac,
    pub adc: StatsAdc,
    pub crc_error_count: AtomicU64,
}

impl StatsAdc {
    pub fn update_values(&self, _values: AdcPoint) {
        unimplemented!()
    }
}

/// Number of DAC points to write until notified.
// FIXME: Adjust value.
const DAC_NOTIFY_EVERY: usize = 100;
/// Number of ADC points to read until notified.
// FIXME: Adjust value.
const ADC_NOTIFY_EVERY: usize = 100;

pub struct ControlHandle {
    /// Semaphore to notify that something is ready.
    ready_sem: Semaphore,

    dac_enabled: AtomicBool,

    din: AtomicDin,
    dout: AtomicDout,

    /// Discrete input has changed.
    din_changed: AtomicBool,
    /// Discrete output has changed.
    dout_changed: AtomicBool,
}

struct ControlDac {
    running: bool,
    buffer: DacConsumer,
    last_point: DacPoint,
    counter: usize,
}

struct ControlAdc {
    buffer: AdcProducer,
    last_point: AdcPoint,
    counter: usize,
}

pub struct Control {
    dac: ControlDac,
    adc: ControlAdc,
    handle: Arc<ControlHandle>,
}

impl ControlHandle {
    fn new() -> Self {
        Self {
            ready_sem: Semaphore::new_binary().unwrap(),
            dac_enabled: AtomicBool::new(false),
            din: AtomicDin::new(0),
            dout: AtomicDout::new(0),
            din_changed: AtomicBool::new(false),
            dout_changed: AtomicBool::new(false),
        }
    }

    fn update_din(&self, din: Din) -> bool {
        if self.din.swap(din, Ordering::AcqRel) != din {
            self.din_changed.fetch_or(true, Ordering::AcqRel);
            true
        } else {
            false
        }
    }
}

impl Control {
    pub fn new(dac_buf: DacConsumer, adc_buf: AdcProducer) -> (Self, Arc<ControlHandle>) {
        let handle = Arc::new(ControlHandle::new());
        (
            Self {
                dac: ControlDac {
                    running: false,
                    buffer: dac_buf,
                    last_point: 0x7fff,
                    counter: 0,
                },
                adc: ControlAdc {
                    buffer: adc_buf,
                    last_point: AdcPoint::default(),
                    counter: 0,
                },
                handle: handle.clone(),
            },
            handle,
        )
    }

    fn make_din_handler(&self) -> impl DinHandler {
        let handle = self.handle.clone();
        move |context, din| {
            if handle.update_din(din) {
                handle.ready_sem.give_from_isr(context);
            }
        }
    }

    fn task_main(&mut self, stats: Arc<Statistics>) {
        let mut skifio = skifio::handle().unwrap();
        skifio.subscribe_din(Some(self.make_din_handler())).unwrap();

        println!("Enter SkifIO loop");
        //uint64_t prev_intr_count = _SKIFIO_DEBUG_INFO.intr_count;
        let iter_counter = 0;
        loop {
            let mut ready = false;

            skifio
                .set_dac_state(self.handle.dac_enabled.load(Ordering::Acquire))
                .unwrap();

            // Wait for 10 kHz sync signal
            match skifio.wait_ready(Some(Duration::from_millis(1000))) {
                Ok(()) => (),
                Err(Error::Hal(RetCode::TimedOut)) => {
                    println!("SkifIO timeout {}", iter_counter);
                    continue;
                }
                Err(e) => panic!("{:?}", e),
            }

            // Write discrete output
            if self.handle.dout_changed.fetch_and(false, Ordering::AcqRel) {
                skifio
                    .write_dout(self.handle.dout.load(Ordering::Acquire))
                    .unwrap();
            }

            // Read discrete input
            ready |= self.handle.update_din(skifio.read_din());

            // Statistics: detect 10 kHz sync signal loss
            /*
            self->stats->max_intrs_per_sample = hal_max(
                self->stats->max_intrs_per_sample,
                (uint32_t)(_SKIFIO_DEBUG_INFO.intr_count - prev_intr_count) //
            );
            prev_intr_count = _SKIFIO_DEBUG_INFO.intr_count;
            */

            // Fetch next DAC value from buffer
            let mut dac_value = self.dac.last_point;
            if self.dac.running {
                if let Some(value) = self.dac.buffer.pop() {
                    dac_value = value;
                    self.dac.last_point = value;
                    // Decrement DAC notification counter.
                    if self.dac.counter > 0 {
                        self.dac.counter -= 1;
                    } else {
                        self.dac.counter = DAC_NOTIFY_EVERY - 1;
                        ready = true;
                    }
                } else {
                    stats.dac.lost_empty.fetch_add(1, Ordering::AcqRel);
                }
            }

            // Transfer DAC/ADC values to/from SkifIO board.
            {
                // TODO: Check for overflow.
                let dac = dac_value as Aout;
                let adcs = match skifio.transfer(XferOut { dac }) {
                    Ok(XferIn { adcs }) => {
                        self.adc.last_point = adcs;
                        adcs
                    }
                    Err(Error::Hal(RetCode::InvalidData)) => {
                        // CRC check error
                        stats.crc_error_count.fetch_add(1, Ordering::AcqRel);
                        self.adc.last_point
                    }
                    Err(e) => panic!("{:?}", e),
                };

                // Handle ADCs
                {
                    // Update ADC value statistics
                    stats.adc.update_values(adcs);
                    // Push ADC point to buffer.
                    if self.adc.buffer.push(adcs).is_err() {
                        stats.adc.lost_full.fetch_add(1, Ordering::AcqRel);
                    }

                    // Decrement ADC notification counter.
                    if self.adc.counter > 0 {
                        self.adc.counter -= 1;
                    } else {
                        self.adc.counter = ADC_NOTIFY_EVERY - 1;
                        ready = true;
                    }
                }
            }

            if ready {
                // Notify
                self.handle.ready_sem.give();
            }

            stats.sample_count.fetch_add(1, Ordering::AcqRel);
        }
    }

    pub fn run(mut self, priority: u8, stats: Arc<Statistics>) -> Result<Task, Error> {
        Task::new()
            .name("control")
            .priority(TaskPriority(priority))
            .start(move |_| self.task_main(stats))
            .map_err(|e| e.into())
    }
}
